diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
index 71b247c2315..6e7f0678c6f 100644
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -2724,4 +2724,6 @@
          is installed. -->
     <bool name="config_permissionReviewRequired">false</bool>
 
+    <!-- Whether to enable HumanInteractionController by default -->
+    <bool name="config_HICEnabledDefault">true</bool>
 </resources>
diff --git a/core/res/res/values/symbols.xml b/core/res/res/values/symbols.xml
index 3a762207efd..35437347b1c 100644
--- a/core/res/res/values/symbols.xml
+++ b/core/res/res/values/symbols.xml
@@ -2751,4 +2751,7 @@
   <java-symbol type="bool" name="use_lock_pattern_drawable" />
   <java-symbol type="drawable" name="lockscreen_notselected" />
   <java-symbol type="drawable" name="lockscreen_selected" />
+
+  <!-- Whether to enable HumanInteractionController by default -->
+  <java-symbol type="bool" name="config_HICEnabledDefault" />
 </resources>
diff --git a/libs/hwui/Android.mk b/libs/hwui/Android.mk
index cacfce16ab9..c54feeb260c 100644
--- a/libs/hwui/Android.mk
+++ b/libs/hwui/Android.mk
@@ -149,7 +149,7 @@ endif
 
 ifdef HWUI_COMPILE_FOR_PERF
     # TODO: Non-arm?
-    hwui_cflags += -fno-omit-frame-pointer -marm -mapcs
+    hwui_cflags += -fno-omit-frame-pointer -marm
 endif
 
 # This has to be lazy-resolved because it depends on the LOCAL_MODULE_CLASS
diff --git a/packages/MtpDocumentsProvider/src/com/android/mtp/MtpDocumentsProvider.java b/packages/MtpDocumentsProvider/src/com/android/mtp/MtpDocumentsProvider.java
index f1e24b80fad..8b59a578d36 100644
--- a/packages/MtpDocumentsProvider/src/com/android/mtp/MtpDocumentsProvider.java
+++ b/packages/MtpDocumentsProvider/src/com/android/mtp/MtpDocumentsProvider.java
@@ -137,6 +137,9 @@ public class MtpDocumentsProvider extends DocumentsProvider {
         } catch (IOException error) {
             Log.e(TAG, "Failed to start app fuse.", error);
             return false;
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to mount with appfuse.", e);
+            return false;
         }
 
         resume();
diff --git a/packages/SystemUI/src/com/android/systemui/classifier/HumanInteractionClassifier.java b/packages/SystemUI/src/com/android/systemui/classifier/HumanInteractionClassifier.java
index 851ab771882..fe615f9105c 100644
--- a/packages/SystemUI/src/com/android/systemui/classifier/HumanInteractionClassifier.java
+++ b/packages/SystemUI/src/com/android/systemui/classifier/HumanInteractionClassifier.java
@@ -36,9 +36,6 @@ public class HumanInteractionClassifier extends Classifier {
     private static final String HIC_ENABLE = "HIC_enable";
     private static final float FINGER_DISTANCE = 0.1f;
 
-    /** Default value for the HIC_ENABLE setting: 1 - enabled, 0 - disabled */
-    private static final int HIC_ENABLE_DEFAULT = 1;
-
     private static HumanInteractionClassifier sInstance = null;
 
     private final Handler mHandler = new Handler();
@@ -105,9 +102,11 @@ public class HumanInteractionClassifier extends Classifier {
     }
 
     private void updateConfiguration() {
+        boolean enabledDefault = mContext.getResources()
+                .getBoolean(com.android.internal.R.bool.config_HICEnabledDefault);
         mEnableClassifier = 0 != Settings.Global.getInt(
                 mContext.getContentResolver(),
-                HIC_ENABLE, HIC_ENABLE_DEFAULT);
+                HIC_ENABLE, enabledDefault ? 1 : 0);
     }
 
     public void setType(int type) {
